src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesSlave.java

```diff
@@ -178,10 +183,40 @@ public KubernetesComputer createComputer() {
         return new KubernetesComputer(this);
     }
 
+    public PodRetention getPodRetention(KubernetesCloud cloud) {
+        PodRetention retentionPolicy = cloud.getPodRetention();
+        if (template != null && template.getPodRetention() != PodRetention.DEFAULT) {
+            retentionPolicy = template.getPodRetention();
+        }
+        return retentionPolicy;
+    }
+
     @Override
     protected void _terminate(TaskListener listener) throws IOException, InterruptedException {
         LOGGER.log(Level.INFO, "Terminating Kubernetes instance for agent {0}", name);
+        
+        KubernetesCloud cloud;
+        try {
+            cloud = getKubernetesCloud();
+        } catch (IllegalStateException e) {
+            e.printStackTrace(listener.fatalError("Unable to terminate agent. Cloud may have been removed. There may be leftover resources on the Kubernetes cluster."));
+            LOGGER.log(Level.SEVERE, String.format("Unable to terminate agent %s. Cloud may have been removed. There may be leftover resources on the Kubernetes cluster.", name));
+            return;
+        }
+
+        KubernetesClient client;
+        try {
+            client = cloud.connect();
+        } catch (UnrecoverableKeyException | CertificateEncodingException | NoSuchAlgorithmException
+                | KeyStoreException e) {
+            String msg = String.format("Failed to connect to cloud %s. There may be leftover resources on the Kubernetes cluster.", getCloudName());
+            e.printStackTrace(listener.fatalError(msg));
+            LOGGER.log(Level.SEVERE, msg);
+            return;
+        }
 
+        boolean deletePod = shouldDeletePod(client, cloud);
```

Comment:
Check if we should delete the pod _before_ disconnecting the JNLP agent. At this point a healthy build will have a running pod with a JNLP agent running.

Reply:
A code comment here seems warranted